benchclients/python/benchclients/conbench.py

```diff
@@ -1,42 +1,152 @@
+        The email address to use for Conbench login. Required for submitting
+        data.
     CONBENCH_PASSWORD
-        The password to use for Conbench login. Only required for GETting if the server is private.
+        The password to use for Conbench login. Required for submitting data.
+
+    Credentials can be left undefined when only reading state from a 'public
+    mode' API server.
     """
 
-    def __init__(self, adapter: Optional[HTTPAdapter] = None):
+    default_retry_for_seconds = 3 * 60
+    timeout_login_request = (10, 3.5)
+    timeout_long_running_requests = (120, 3.5)
+
+    def __init__(self, default_retry_for_seconds=None, adapter=None):
+        # If this library is embedded into a Python program that has stdlib
+        # logging not set up yet (no root logger configured) then this call
+        # sets up a root logger with handlers. This is a noop if the calling
+        # program has a root logger already configured.
+        logging.basicConfig(
+            level=logging.INFO,
+            format="%(asctime)s.%(msecs)03d [%(name)s] %(levelname)s: %(message)s",
+            datefmt="%y%m%d-%H:%M:%S",
+        )
+
+        url = self._read_base_url_from_env_or_raise()
+
+        if adapter:
+            # Let's phase this out.
+            log.info("ignoring adapter: %s", adapter)
+
+        super().__init__()
+
+        if default_retry_for_seconds:
+            assert isinstance(default_retry_for_seconds, (float, int))
+            self.default_retry_for_seconds = default_retry_for_seconds
+
+        # Construct the HTTP API base URL without a trailing slash, something
+        # like https://conbench.ursa.dev/api
+        self._url = url.rstrip("/") + "/api"
+        # self.default_retry_for_seconds = default_retry_for_seconds
+
+        if "CONBENCH_EMAIL" in os.environ:
+            # The logic would attempt to perform login automatically after
+            # receiving the first 401 response. When this env var is set,
+            # anticipate that login is needed (this might do more harm than
+            # use)
+            self._login_or_raise()
+        else:
+            log.info("CONBENCH_EMAIL not in environment, skipping login")
+
+        log.info("%s: initialized", self.__class__.__name__)
+
+    # This method is required by parent class
+    @property
+    def _base_url(self) -> str:
+        return self._url
+
+    def _read_base_url_from_env_or_raise(self) -> str:
+        # Maybe rename to CONBENCH_BASE_URL.
         url = os.getenv("CONBENCH_URL")
+        if url:
+            # Strip leading and trailing whitespace
+            url = url.strip()
+
         if not url:
-            fatal_and_log("Environment variable CONBENCH_URL not found")
+            raise ConbenchClientException(
+                "Environment variable CONBENCH_URL not set or empty"
+            )
+
+        # Try to catch a common user error.
+        if url.endswith("/api"):
```

Comment:
Want to check for `/api/` too?

Reply:
thanks, yes, also cleaning this up a bit